pilot/pkg/xds/mtls_checker.go

```diff
@@ -15,135 +15,80 @@
-		peerAuthDisabledMTLS: map[string]bool{},
-		subsetPolicyMode:     map[string]*networkingapi.ClientTLSSettings_TLSmode{},
-		rootPolicyMode:       tlsModeForDefaultTrafficPolicy(dr, svcPort),
+		push:            push,
+		svcPort:         svcPort,
+		destinationRule: tlsModeForDestinationRule(dr, subset, svcPort),
 	}
 }
 
-// isMtlsDisabled returns true if the given lbEp has mTLS disabled due to any of:
-// - disabled tlsMode
-// - DestinationRule disabling mTLS on the entire host or the port
-// - PeerAuthentication disabling mTLS at any applicable level (mesh, ns, workload, port)
+// isMtlsDisabled returns true if the given lbEp has mTLS disabled due to any of
+// This must be called after computeMtlsEnabled
 func (c *mtlsChecker) isMtlsDisabled(lbEp *endpoint.LbEndpoint) bool {
-	if c == nil {
-		return false
-	}
-	_, ok := c.mtlsDisabledHosts[lbEpKey(lbEp)]
-	return ok
+	return lbEp.Metadata.FilterMetadata[util.EnvoyTransportSocketMetadataKey].
+		GetFields()[model.TLSModeLabelShortname].
+		GetStringValue() != model.IstioMutualTLSModeLabel
 }
 
-// computeForEndpoint checks destination rule, peer authentication and tls mode labels to determine if mTLS was turned off.
-func (c *mtlsChecker) computeForEndpoint(ep *model.IstioEndpoint) {
-	if drMode := c.tlsModeForDestinationRule(ep); drMode != nil {
+// computeMtlsEnabled computes whether mTLS should be enabled or not
+func (c *mtlsChecker) computeMtlsEnabled(ep *model.IstioEndpoint) bool {
+	if drMode := c.destinationRule; drMode != nil {
 		switch *drMode {
 		case networkingapi.ClientTLSSettings_DISABLE:
-			c.mtlsDisabledHosts[lbEpKey(ep.EnvoyEndpoint)] = struct{}{}
-			return
+			return false
 		case networkingapi.ClientTLSSettings_ISTIO_MUTUAL:
 			// don't mark this EP disabled, even if PA or tlsMode meta mark disabled
-			return
+			return true
 		}
 	}
 
 	// if endpoint has no sidecar or explicitly tls disabled by "security.istio.io/tlsMode" label.
 	if ep.TLSMode != model.IstioMutualTLSModeLabel {
-		c.mtlsDisabledHosts[lbEpKey(ep.EnvoyEndpoint)] = struct{}{}
-		return
-	}
-
-	mtlsDisabledByPeerAuthentication := func(ep *model.IstioEndpoint) bool {
-		// apply any matching peer authentications
-		peerAuthnKey := ep.Labels.String() + ":" + strconv.Itoa(int(ep.EndpointPort))
-		if value, ok := c.peerAuthDisabledMTLS[peerAuthnKey]; ok {
-			// avoid recomputing since most EPs will have the same labels/port
-			return value
-		}
-		c.peerAuthDisabledMTLS[peerAuthnKey] = factory.
-			NewMtlsPolicy(c.push, ep.Namespace, ep.Labels).
-			GetMutualTLSModeForPort(ep.EndpointPort) == model.MTLSDisable
-		return c.peerAuthDisabledMTLS[peerAuthnKey]
+		return false
 	}
 
-	//  mtls disabled by PeerAuthentication
-	if mtlsDisabledByPeerAuthentication(ep) {
-		c.mtlsDisabledHosts[lbEpKey(ep.EnvoyEndpoint)] = struct{}{}
-	}
+	return factory.
+		NewMtlsPolicy(c.push, ep.Namespace, ep.Labels).
+		GetMutualTLSModeForPort(ep.EndpointPort) != model.MTLSDisable
```

Comment:
Not sure how the perf was improved when we removed the PA policy cache. I think it is because the benchmark does not include PA

Reply:
Good call, I expanded the benchmark so we do not have to rely on manual testing in real world to reproduce.

The cache key is (orders of magnitude) more expensive than the PA computation, so counter-intuitively it actually is faster to do this than to use the cache.

Here is the new benchmark results:
```
name                          old time/op        new time/op        delta
EndpointGeneration/1/100-48         1.04ms ± 1%        0.78ms ± 1%  -24.77%  (p=0.000 n=10+10)
EndpointGeneration/10/10-48          610µs ± 1%         444µs ± 1%  -27.29%  (p=0.000 n=9+9)
EndpointGeneration/100/10-48        5.70ms ± 3%        4.15ms ± 3%  -27.12%  (p=0.000 n=10+10)
EndpointGeneration/1000/1-48        5.61ms ± 3%        4.06ms ± 3%  -27.74%  (p=0.000 n=10+9)

name                          old kb/msg         new kb/msg         delta
EndpointGeneration/1/100-48           16.2 ± 0%          16.2 ± 0%     ~     (all equal)
EndpointGeneration/10/10-48           13.0 ± 0%          13.0 ± 0%     ~     (all equal)
EndpointGeneration/100/10-48           128 ± 0%           128 ± 0%     ~     (all equal)
EndpointGeneration/1000/1-48           129 ± 0%           129 ± 0%     ~     (all equal)

name                          old resources/msg  new resources/msg  delta
EndpointGeneration/1/100-48            100 ± 0%           100 ± 0%     ~     (all equal)
EndpointGeneration/10/10-48           10.0 ± 0%          10.0 ± 0%     ~     (all equal)
EndpointGeneration/100/10-48          10.0 ± 0%          10.0 ± 0%     ~     (all equal)
EndpointGeneration/1000/1-48          1.00 ± 0%          1.00 ± 0%     ~     (all equal)

name                          old alloc/op       new alloc/op       delta
EndpointGeneration/1/100-48          364kB ± 0%         234kB ± 0%  -35.63%  (p=0.000 n=10+10)
EndpointGeneration/10/10-48          204kB ± 0%         122kB ± 0%  -40.21%  (p=0.000 n=9+9)
EndpointGeneration/100/10-48        1.78MB ± 0%        1.09MB ± 0%  -38.74%  (p=0.000 n=10+8)
EndpointGeneration/1000/1-48        1.76MB ± 0%        1.09MB ± 0%  -38.34%  (p=0.000 n=9+9)

name                          old allocs/op      new allocs/op      delta
EndpointGeneration/1/100-48          7.91k ± 0%         5.21k ± 0%  -34.12%  (p=0.000 n=10+10)
EndpointGeneration/10/10-48          4.61k ± 0%         2.81k ± 0%  -39.05%  (p=0.000 n=10+10)
EndpointGeneration/100/10-48         39.0k ± 0%         24.6k ± 0%  -36.99%  (p=0.000 n=10+9)
EndpointGeneration/1000/1-48         38.2k ± 0%         24.2k ± 0%  -36.81%  (p=0.000 n=7+9)
```

Which actually shows bigger improvement than the original

Flame before: https://flamegraph.com/share/8ef4cefb-008b-11ee-ba99-de9431916b05

Flame after: https://flamegraph.com/share/7e11da29-008b-11ee-9e85-32d05458fc35

Reply:
Also one thing to note is most of the CPU is spent on protobuf marshal and GC, so 25% speedup is really much larger speedup for the main logic

Reply:
Great tests